README.md
# Cov-Aid Bot

Hello World, this is the repository for **CovAid** Bot. With mission to provide accurate information during COVID-19 Pandemic. This is a Discord bot that you can install to your server for free. 

## Getting Started

### Add to your own discord server
To install this Discord bot into your own Discord server, [click on this link](https://discordapp.com/oauth2/authorize?&client_id=701186191940255785&scope=bot&permissions=55296) to grant the authroization. You must be an administrator in order to add this bot. 

### Run this bot on your own bot.
First is clone this GitHub Repo:
```bash
git clone https://github.com/CovAid-App/CovAid.git
cd CovAid
```
then open `discord.py` and leave it aside for now.

First, make sure you have an discord account.

Then, go to [Discord Developer](https://discordapp.com/developers/applications), then click on "New Applications". Type a name and click "Create". Click in to the application, copy the client ID. Now click on the bot tab at the right side, click on "Create a Bot". Go ahead go to https://discordapp.com/oauth2/authorize?&client_id=[YOUR CLIENT ID]&scope=bot&permissions=55296 to add the discord bot to your own server. Navigate back to the bot tab, and copy the token.

Finally, navigate back `discord.py`, paste your token at the last line ( `client.run("YOUR TOKEN HERE")`).

To start this bot, you need to make sure you have `python3` and `pip` installed. (To install python, go to this [website](https://python.org), `pip` should come with `python`.) Then run following commands:
```bash
pip3 install -r requirements.txt
python3 discord_bot.py
```

First, in terminal, you should see the some green text and red text, that's the accuracy for the bot. Give it about 1 minutes or so to train the chatbot. And the bot should come online. 

### Use the bot
To use the bot, make sure to tag the bot (@CovAidBot), then type your question. You should receive the response within 10 seconds. If not, there might be some error happening. You can file an issue.

### Tests
To see all the tests available, you can go to `test.py`.

### Demo

## Features
1. **Just Chattin'** - You can ask some basic question about COVID-19, such as what is covid-19.
2. **Country Data** - You can search for data in country. You must include country name and one of following keywords: NewConfirmed, TotalConfirmed, NewDeaths, TotalDeaths, NewRecovered, or TotalRecovered.
3. **tips** -  Type `tips` you can see Public Announcement image made by Google/WHO about how to avoid getting COVID-19.
4. **Where are the grocery near me?** - To search for the groceries near an address, you can type `grocery-<YOUR CITY NAME/ADDRESS>`. (Your Address will only be use for search grocery.)
5. **When is the best time to visit grocery?** - To check the best time to visit a grocery, you can type `grocery-<YOUR CITY NAME/ADDRESS>-<GROCERY NAME>`.
6. **Detect whether you have COVID-19?** - We also offered a few links to self-detect whether you have COVID-19 or not. Type detect to get the links.


## Deployment

To Deploy this Discord bot, first see "Run this bot on your own bot".
Then go to [Heroku](https://heroku.com), login or register. Then create a new app, it's up to you the name for the app. Then make sure you have [Heroku CLI](https://devcenter.heroku.com/articles/heroku-cli#download-and-install) installed. Next, head back to your Terminal, then execute following command:
```bash
heroku login
heroku git:remote -a [YOUR APP NAME]
git add .
git commit -m "initial commit"
git push heroku master
```
Wait for about 1 or 2 minutes since there's a lot of packages to install. (trust me, a lot.) After it commited, go to [Heroku](https://heroku.com), open up your app, then click on the "Resource" tab on the top, you should see `worker python discord_bot.py`, click on the pencil button on the right, then toggle, click confirm. 

And now go to your server where your bot added, and about 1 minute or so, the bot should come back online, and you can ask your bot question!

## Built With

* [Discord.py](https://discordpy.readthedocs.io/en/latest/) - Discord Python Library
* [Google Maps API](https://developers.google.com/maps/documentation) - Dependency Management

## Contributing

Please read contribute.md for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/CovAid-App/CovAid/tags). 

## Authors

* **@boyuan12** - *Basic Chatbot, Country API, and tips features and popular time* - [boyuan12](https://github.com/boyuan12)
* **@GamingDevilsCC** - *String for detect* - [GamingDevilsCC](https://github.com/GamingDevilsCC)
* **@thafiona13** - *Worked on logo for this project and suggestobot* - [thafiona13](https://github.com/thafiona13)
* **@bornasadeghi** - *Worked on grocery w/Google Map API and popular time* - [bornasadeghi](https://github.com/bornasadeghi)

See also the list of [contributors](https://github.com/CovAid-App/CovAid/contributors) who participated in this project.

## License

This project is licensed under the MIT License - see the LICENSE file for details


## Acknowledgments

* Please, humanity, learn the lesson from nature, this disease won't just leave us, but let's work together to solve. Our planet is in danger, because we all know why this pandemic happened in the first place. And please, NASA, don't try to find another planet, because we will just destroy that planet like we did to Earth.

changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2020-05-03
### Added
- Initial release
- Include features:
    1. **Just Chattin'** - You can ask some basic question about COVID-19, such as what is covid-19.
    2. **Country Data** - You can search for data in country. You must include country name and one of following keywords: NewConfirmed, TotalConfirmed, NewDeaths, TotalDeaths, NewRecovered, or TotalRecovered.
    3. **tips** -  Type `tips` you can see Public Announcement image made by Google/WHO about how to avoid getting COVID-19.
    4. **Where are the grocery near me?** - To search for the groceries near an address, you can type `grocery-<YOUR CITY NAME/ADDRESS>`. (Your Address will only be use for search grocery.)
    5. **When is the best time to visit grocery?** - To check the best time to visit a grocery, you can type `grocery-<YOUR CITY NAME/ADDRESS>-<GROCERY NAME>`.
    6. **Detect whether you have COVID-19?** - We also offered a few links to self-detect whether you have COVID-19 or not. Type detect to get the links.



test.py
from bot import get_response
import random

def test_bot(tests, times=100):
    """
    Get response from the chatbot

    Args:
        message (str): the message that you want to get the response

    Returns:
        answer (str): the message responded from the chatbot
    """
    correct = 0

    test = []
    response = []

    for key, value in tests.items():
        test.append(key)
        response.append(value)

    total = len(test)

    print(f'Test running for {times} times')
    for i in range(times):
        for j in range(total):
            val = random.choice(list(tests.items()))
            if str(get_response(val[0])) == val[1]:
                correct += 1

    print('Test finished')

    return correct, total * times

trackobot/test.py
import trackobot
from datetime import datetime

user_coords = trackobot.get_user_coordinates()

query = input("Where do you want to go? ")

# take the first ten results
results = trackobot.maps_search(query)[:10]
distances_info = trackobot.get_distances_info([user_coords for each in results], [trackobot.get_place_coordinates(place) for place in results])

for place_index in range(len(results)):
    print(trackobot.get_distance(distances_info, place_index))

'''

for i in range (len(results)):
    place = results[i]
    location = place['geometry']['location']
    lat,lng = location['lat'], location['lng']
    # TODO check if the method allows multiple distance calculations per api call
    # hide distance for now to prevent overdoing api calls
    # distance = trackobot.get_distance(user_coords, (lat,lng))

    print("%-2d - %s (%s)"%(i,place['name'], "dist"))

place_index = int(input("Please choose a place (0 - %d): " % (len(results)-1) ))
selected_place = results[place_index]
place_id = selected_place['place_id']



popular_times_info = trackobot.get_popular_times_info(place_id)

# 0: Monday, up to 6: Sunday
weekday_index = datetime.today().weekday()

print(trackobot.popular_times_on_weekday(popular_times_info, weekday_index))
'''

web.py
from flask import Flask, render_template
import datetime
import pycountry
import requests
import json

from termcolor import colored

app = Flask(__name__)

@app.route("/")
def index():
    pass


@app.route("/track")
def trackobot():
    """Top 4 countries that have the most cases"""
    render_template("loading.html")
    today = datetime.date.today()
    day_7 = today - datetime.timedelta(days=7)
    day_1 = today - datetime.timedelta(days=1)

    country_data = {}
    top = []
    alpha2_to_slug = {}

    data = requests.get("https://coronavirus-19-api.herokuapp.com/countries").json()

    for i in range(1, 6):
        top.append(data[i]["country"])

    for i in (list(pycountry.countries)):
        for j in top:
            if j == i.name or j == i.alpha_3:
                alpha2_to_slug[i.alpha_2] = None

    co = requests.get("https://api.covid19api.com/countries").json()
    for key, value in alpha2_to_slug.items():
        for c in co:
            if c['ISO2'] == key:
                alpha2_to_slug[c["ISO2"]] = c["Slug"]

    current = requests.get("https://api.covid19api.com/summary").json()["Countries"]

    for key, value in alpha2_to_slug.items():
        info = requests.get(f"https://api.covid19api.com/country/{value}/status/confirmed?from={day_7}T00:00:00Z&to={day_1}T00:00:00Z").json()
        cases_avg = 0
        for i in info:
            cases_avg += i["Cases"]

        cases_avg /= 7

        for c in current:
            if c["NewConfirmed"] > cases_avg:
                country_data[value] = False
            else:
                country_data[value] = True

    # {slug: True/False flatten the curve}
    # return str(country_data)
    return render_template("track.html")


@app.route("/chat")
def chat():
    pass


@app.route("/suggest")
def suggest():
    pass


@app.route("/info")
def info():
    pass



track.py
import googlemaps, populartimes

api_key = "AIzaSyAIe4GOG3T6JA5QexlfJKTSP5QSsCckg-I"
client = googlemaps.Client(key=api_key)
print(client.geolocate()['location'])

"""def get_current_coordinates():
    location = client.geolocate()['location']
    lat,lng = location['lat'], location['lng']
    return lat,lng"""

def get_info(lat, lng):

    info = client.places("grocery store", (lat, lng))
    return info

def best_time(info, place_index, day):

    place_id = info["results"][place_index]["place_id"]

    days = {"Monday": 0, "Tuesday": 1, "Wednesday": 2, "Thursday": 3, "Friday": 4, "Saturday": 5, "Sunday": 6}
    data = populartimes.get_id(api_key, place_id)["populartimes"][days[day]]["data"]

    min_val = float('Inf')
    time = 0
    for i in range(len(data)):
        if data[i] != 0 and data[i] < min_val:
            min_val = data[i]
            time = i
    return time


def get_grocery(lat, lng):

    info = get_info(lat, lng)

    result_dict = {}

    for i in range (len(info['results'])):
        place = info['results'][i]
        result_dict[place['name']] = i

    return result_dict, info

discord_bot.py
import discord
from bot import get_response, country_data
from test import test_bot
from termcolor import colored
from track import get_grocery, best_time
import os
from datetime import datetime as date
import requests
from detect import test_string

# start a new discord client
client = discord.Client()

# run the test first
tests = {'what is covid-19?': 'Coronavirus disease 2019 (COVID-19) is an infectious disease caused by severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2). -- Wikipedia'}
correct, total = test_bot(tests)

correct = round(correct / total, 2)
print(colored('Correct: ' + str(correct) + '%', 'green'))
print(colored('Incorrect: ' + str(round(1 - correct, 2)) + '%', 'red'))

day = date.today().strftime("%A")

# print in console of bot name
@client.event
async def on_ready():
    print('We have logged in as {0.user}'.format(client))

# do when receive a new message
@client.event
async def on_message(message):

    # if tagged the bot
    id = client.user.id
    if str(id) in message.content:

        # get the question
        resp = str(message.content).split("<@!701186191940255785> ")[1]

        # if question is tips
        if resp == "tips":

            # send a message with tips.png
            await message.channel.send('Here are the 5 tips to avoid getting COVID-19', file=discord.File('tips.png'))

        # pass through country_data function and if not False
        elif country_data(resp) != False:

            # get the output from that function and send the message, and send the message
            value, date, country, keyword = country_data(resp)
            await message.channel.send(f'As of {date}, in {country}, {keyword} is {value}.')

        elif 'grocery' in resp:
            import json
            place = str(resp).split("-")[1]
            geo = requests.get(f"https://api.opencagedata.com/geocode/v1/json?q={place}&key=485f61225dd041a9bae03a54cc0ef746").json()["results"][0]["geometry"]
            data, info = get_grocery(geo['lat'], geo['lng'])

            if resp.count('-') == 2:
                store = str(resp).split("-")[2]
                await message.channel.send(f'Best hour to visit: {best_time(info, data[store], day)}')
            elif resp.count('-') < 2:
                await message.channel.send(f"Here are the grocery store that is close to you: {data}. Which one do you like to go? ")

        elif 'help' in resp:
            await message.channel.send(
            """
            Hello World! This is CovAid Bot! To ask me a question, please tag me then ask your question. There are a few feature included:
            1. **Just Chattin'** - You can ask some basic question about COVID-19, such as what is covid-19.
            2. **Country Data** - You can search for data in country. You must include country name and one of following keywords: NewConfirmed, TotalConfirmed, NewDeaths, TotalDeaths, NewRecovered, or TotalRecovered.
            3. **tips** Type `tips` you can see Public Announcement image made by Google/WHO about how to avoid getting COVID-19.
            4. **Where are the grocery near me?** - To search for the groceries near an address, you can type `grocery-<YOUR CITY NAME/ADDRESS>`. (Your Address will only be use for search grocery.)
            5. **When is the best time to visit grocery?** - To check the best time to visit a grocery, you can type `grocery-<YOUR CITY NAME/ADDRESS>-<GROCERY NAME>`.
            6. **Detect whether you have COVID-19?** - We also offered a few links to self-detect whether you have COVID-19 or not. Type `detect` to get the links.
            Thanks for using this bot! We hope you get the info you need, and we will get through this unprecedented time together! :)
            """)

        elif "detect" in resp:
            await message.channel.send(test_string)

        # if it none of these, use regular ChatBot
        else:
            await message.channel.send(str(get_response(resp)))


# return the client using this client id
client.run("NzAxMTg2MTkxOTQwMjU1Nzg1.Xq9Ccg.PN10FDGcbVvblivSgx6zF88W9Jc")

trackobot/trackobot.py
import googlemaps, populartimes

api_key = "AIzaSyAIe4GOG3T6JA5QexlfJKTSP5QSsCckg-I"
client = googlemaps.Client(key=api_key)

def get_user_coordinates():
    location = client.geolocate()['location']
    lat,lng = location['lat'], location['lng']
    return lat,lng

def get_place_coordinates(place):
    return place['geometry']['location']

def get_distances_info(origins, destinations):
    '''

    '''
    dist_info = client.distance_matrix(origins, destinations, "driving")
    
    return dist_info

def get_distance(distances_info, index):
    return distances_info['rows'][index]['elements'][0]['distance']['text']

def get_popular_times_info(place_id):
    '''
    Returns JSON of popular times for the given place.
    '''
    return populartimes.get_id(api_key, place_id)['populartimes']

def popular_times_on_weekday(popular_times_info, weekday_index):
    '''
    Returns the popular times of a place for a certain weekday.
    '''
    return popular_times_info[weekday_index]['data']

def get_best_time_to_visit(popular_times_on_weekday):
    '''
    Gets the hour of minimum popularity for the given weekday.
    '''
    return min(pop for pop in popular_times_on_weekday if pop > 0)

def maps_search(query):
    '''
    Searches for nearby places given a query.
    '''
    info = client.places(query)
    return info['results']

bot.py
import os
try:
    os.remove("db.sqlite3")
    os.remove("sentence_tokenizer.pickle")
except FileNotFoundError:
    pass

from chatterbot import ChatBot
from chatterbot.trainers import ListTrainer
from datetime import date
import json
import requests
import pycountry

# use covid19api to get data
covid19 = requests.request("GET", "https://api.covid19api.com/summary")

# create a new ChatBot using chatterbot
chatbot = ChatBot(
    "CovAid",
    logic_adapters=[
        "chatterbot.logic.BestMatch"
    ]
)

# create a ListTrainer
trainer = ListTrainer(chatbot)

# List of item that need to trained
trainer.train(['covid death global', f'As of {date.today()}, there are {covid19.json()["Global"]["TotalDeaths"]} deaths caused by COVID-19 worldwide'])
trainer.train(['covid confirmed global', f'As of {date.today()}, there are {covid19.json()["Global"]["TotalConfirmed"]} confirmed cases of COVID-19 worldwide.'])
trainer.train(['covid tips', f'1. STAY home as much as you can; 2. KEEP a safe distance; 3. WASH hands often 4. COVER your cough 5. SICK? Call ahead'])
trainer.train(['what is covid-19?', f'Coronavirus disease 2019 (COVID-19) is an infectious disease caused by severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2). -- Wikipedia'])
trainer.train(['definition for covid19', f'Coronavirus disease 2019 (COVID-19) is an infectious disease caused by severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2). -- Wikipedia'])
trainer.train(['info', f'For absolute emergency, please dial **9-1-1** immediately, or your countries\' emergercy hotline. Please visit https://cdc.gov and https://who.int for accurate information.'])
trainer.train(['symptoms', f"Common symptoms: fever. tiredness. dry cough.\nSome people may experience: aches and pains. nasal congestion. runny nose. sore throat. diarrhoea."])
trainer.train(['who are you?', f"I'm CovAid Bot! With responsibility to provide accurate information during this COVID-19 Pandemic. To ask a question, <@!701186191940255785> with your question!"])
trainer.train(['What language are you programmed in?', f"I'm programmed in Python!"])
trainer.train(['donate', f"Thanks for your generosity, and no matter how much did you contribute, your contribution will fully go to COVID-19 fund. Go to this website to donate: https://covid19responsefund.org/"])
trainer.train(['is covid-19 present in animals?', f"According to multiple science research paper, they predict that it come from bat."])


# get response from the chatbot
def get_response(message):
    """
    Get response from the chatbot

    Args:
        message (str): the message that you want to get the response

    Returns:
        answer (str): the message responded from the chatbot
    """
    response = chatbot.get_response(message)
    return response


# get the country data
def country_data(sentence, country=None, keyword=None):
    """
    Check a sentence see if it ask for any country's data

    Args:
        sentence (str): the sentence that you want to check

    Returns:
        bool (False) if sentence doesn't contain data
        Tuple (value, date, country, keyword) if sentence does contain data
    """

    # split the word based on the single-space
    words = sentence.split()

    # setup country, keyword variable for later use
    country = False
    keyword = False

    # loop through the list of country from pycountry.countries
    for i in list(pycountry.countries):

        # see if each individual country name is in the sentence, if so, country variable equal to country code (2)
        if i.name in sentence:
            country = i.alpha_2

    # check if any word matches with keyword
    for word in words:
        if word in ['NewConfirmed', 'TotalConfirmed', 'NewDeaths', 'TotalDeaths', 'NewRecovered', 'TotalRecovered']:
            keyword = word

    # if both country and keyword variable are not False, proceed
    if country != False and keyword != False:

        # get response from covid19api (summary route)
        url = "https://api.covid19api.com/summary"
        response = requests.request("GET", url)

        # loop through each individual countries dict
        for ctry in response.json()['Countries']:

            # if country code matches country
            if ctry['CountryCode'] == country:

                # return info
                return ctry[keyword], ctry['Date'], country, keyword

    # return False by default
    return False
